src/utils/sqs_consumer.ts

```diff
@@ -0,0 +1,91 @@
+import { SQS } from 'aws-sdk';
+import delay from 'delay';
+
+import { ONE_SECOND_MS } from '../constants';
+import { logger } from '../logger';
+
+import { SqsClient } from './sqs_client';
+
+export type MessageHandler = (message: SQS.Types.Message) => Promise<any>;
+
+export class SqsConsumer {
+    private readonly _sqsClient: SqsClient;
+    private readonly _beforeHandle?: () => Promise<boolean>;
+    private readonly _handleMessage: MessageHandler;
+    private readonly _afterHandle?: MessageHandler;
+    private _isConsuming: boolean;
+
+    constructor(params: {
+        sqsClient: SqsClient;
+        beforeHandle?: () => Promise<boolean>;
+        handleMessage: MessageHandler;
+        afterHandle?: MessageHandler;
+    }) {
+        this._sqsClient = params.sqsClient;
+        this._beforeHandle = params.beforeHandle;
+        this._handleMessage = params.handleMessage;
+        this._afterHandle = params.afterHandle;
+        this._isConsuming = false;
+    }
+
+    public stop(): void {
+        this._isConsuming = false;
+    }
+
+    public async consumeAsync(): Promise<void> {
+        if (this._isConsuming) {
+            return;
+        }
+
+        this._isConsuming = true;
+        while (this._isConsuming) {
+            await this.consumeOnceAsync();
+        }
+    }
+
+    public async consumeOnceAsync(): Promise<void> {
+        // Run the before hook
+        if (this._beforeHandle) {
+            let beforeCheck;
+            try {
+                beforeCheck = await this._beforeHandle();
+            } catch (e) {
+                logger.warn(e, 'Error encountered in the preHandle check');
+                return;
+            }
+
+            if (!beforeCheck) {
+                logger.warn('before validation failed');
+                await delay(ONE_SECOND_MS);
+                return;
+            }
+        }
+
+        // Receive message
+        const message = await this._sqsClient.receiveMessageAsync();
+
+        // No message
+        if (message === null) {
+            return;
+        }
+
+        // Handle message
+        try {
+            await this._handleMessage(message);
+        } catch (err) {
```

Comment:
Just curious, how can we differentiate here for kinds of errors that are recoverable (errors where we decide to re-try the message in the future or in another worker. Example: network blip) compard to unexpected errors that we do not recover from (order expired, simulation fails). Are the latter not errors in your opinion?

Reply:
Good question! In the past, I've done the following:
- Create a list of "retryable errors" and "nonretryable errors"
- Catch error. If the type of the error is "retryable", then retry. If not, then not

But I think in our case, we can add that functionality later if we feel that it provides value. I think starting with this stance is pretty pragmatic:
- All errors are retryable
- Except for a certain class of errors that we're expecting. These will terminate the job and put the job in its correct status
- All retries, after a certain number of attempts, will get placed on the DLQ, where we can manually merge them later

That said, we could also add an `onError` handler to the SqsConsumer, which would allow us to implement this logic fairly easily